sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -258,6 +258,26 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
       val outputAttributes = readDataColumns ++ generatedMetadataColumns ++
         partitionColumns ++ constantMetadataColumns
 
+      // Rebind metadata attribute references in filters after the metadata attribute struct has
+      // been flattened. Only data filters can contain metadata references. The rebinding will
+      // categorize the references as either [[FileSourceConstantMetadataAttribute]] or
+      // [[FileSourceGeneratedMetadataAttribute]].
+      def rebindFileSourceMetadataAttributesInFilters(
+          filters: Seq[Expression]): Seq[Expression] =
+        filters.map { filter =>
+          filter.transform {
+            case fieldGetter @ GetStructField(FileSourceMetadataAttribute(_), _, _) =>
```

Comment:
what if the predicate is `func_taking_struct(_metadata)`? shall we replace `_metadata` with `CreateStruct(metadataColumns)`?

Reply:
Good point, I will have to deep dive a bit here. Basically I am not sure if this is actually possible or if a filter referencing metadata in the struct form will error out because the struct is not there yet once the initial filter expression is built

Reply:
I added a test for this and it works